app/models/provider/marketstack.rb

```diff
@@ -0,0 +1,143 @@
+class Provider::Marketstack
+  include Retryable
+
+  def initialize(api_key)
+    @api_key = api_key
+  end
+
+  def fetch_security_prices(ticker:, start_date:, end_date:)
+    prices = paginate("#{base_url}/eod", {
+      symbols: ticker,
+      date_from: start_date.to_s,
+      date_to: end_date.to_s
+    }) do |body|
+      body.dig("data").map do |price|
+        {
+          date: price["date"],
+          price: price["close"]&.to_f,
+          currency: "USD"
+        }
+      end
+    end
+
+    SecurityPriceResponse.new(
+      prices: prices,
+      success?: true,
+      raw_response: prices.to_json
+    )
+  rescue StandardError => error
+    SecurityPriceResponse.new(
+      success?: false,
+      error: error,
+      raw_response: error
+    )
+  end
+
+  def fetch_all_tickers
```

Comment:
Looks like we might be able to consolidate into a `fetch_tickers(exchange_mic: nil)` method that covers both `fetch_all_tickers` and `fetch_exchange_tickers`?

```rb
def fetch_tickers(exchange_mic: nil)
  endpoint = "#{base_url}/tickers"
  endpoint = endpoint + "?exchange=#{exchange_mic}" if exchange_mic.present?

  tickers = paginate(endpoint) do |body|
    # ...
  end
end
```

And then over in importer, would reduce to:

```rb
securities = @provider.fetch_tickers(exchange_mic: @exchange)&.tickers
```

Reply:
Good catch1